runtime-modules/proposals/engine/src/lib.rs

```diff
@@ -348,8 +373,57 @@ decl_module! {
         const MaxActiveProposalLimit: u32 = T::MaxActiveProposalLimit::get();
 
+        /// Block Initialization. Perform voting period check, vote result tally, approved proposals
+        /// grace period checks, and proposal execution.
+        fn on_initialize() -> Weight {
+            // `process_proposal` returns the weight of the executed proposals. The weight of the
+            // executed proposals doesn't include any access to the store or calculation that
+            // `on_initialize` does. Therefore, to get the total weight of `on_initialize` we need
+            // to add the weight of the execution of `on_intialize` to the weight returned by
+            // `process_proposal`.
+            // To be safe, we use the worst possible case for `on_initialize`, meaning that there
+            // are as many proposals active as possible and they all take the worst possible branch.
+
+            let max_active_proposals = T::MaxActiveProposalLimit::get();
+
+            // Weight when all the proposals are immediatly approved and executed
+            let immediate_execution_branch_weight =
+                WeightInfoEngine::<T>::
+                on_initialize_immediate_execution_decode_fails(max_active_proposals);
+
+            let pending_execution_branch_weight =
+                WeightInfoEngine::<T>::
+                on_initialize_pending_execution_decode_fails(max_active_proposals);
+
+            // Weight when all the proposals are approved and pending constitutionality
+            let approved_pending_constitutionality_branch_weight =
+                WeightInfoEngine::<T>::
+                on_initialize_approved_pending_constitutionality(max_active_proposals);
+
+            // Weight when all proposals are rejected
+            let rejected_branch_weight =
+                WeightInfoEngine::<T>::on_initialize_rejected(max_active_proposals);
+
+            // Weight when all proposals are slashed
+            let slashed_branch_weight =
+                WeightInfoEngine::<T>::on_initialize_slashed(max_active_proposals);
+
+            // Weight of the executed proposals
+            let executed_proposals_weight = Self::process_proposals();
+
+            // Maximum Weight of all possible worst case scenarios
+            let maximum_branch_weight = immediate_execution_branch_weight
+                .max(pending_execution_branch_weight)
+                .max(approved_pending_constitutionality_branch_weight)
+                .max(rejected_branch_weight)
+                .max(slashed_branch_weight);
+
+            // total_weight = executed_proposals_weight + maximum_branch_weight
+            executed_proposals_weight.saturating_add(maximum_branch_weight)
+        }
+
         /// Vote extrinsic. Conditions:  origin must allow votes.
-        #[weight = 10_000_000] // TODO: adjust weight
+        #[weight = WeightInfoEngine::<T>::vote(_rationale.len().try_into().unwrap())]
```

Comment:
unwrap() and expect() are dangerous outside of the testing context. Please try to replace them with non-panicking alternatives like saturated_into() or another one.